### ISSUE TITLE
Add function to extract username and password from credential embedded URL in url module

### ISSUE THREAD

**Issue by hanlins** [k14s/ytt#366]
**Describe the problem/challenge you have**
[A description of the current challenge that you are experiencing.]
When handling a credential embedded URL in the form of `https://username:password@URL`, we want to parse the credential information like `username` and `password`, store them into a k8s secret. Also we want to get a URL that doesn't contain the credential so we can store it elsewhere.

**Describe the solution you'd like**
[A clear and concise description of what you want to happen. If applicable a visual representation of the UX (ex: new CLI argument name, or resulting yaml).]

Involve following methods in `url` module:
1. `extract_username`: takes in a URL string, returns the embedded username as a string. If there's no credential, return `""`.
2. `extract_password`: takes in a URL string, returns the embedded password as a string. If there's no credential, return `""`.
3. `trim_credentials`: takes in a URL string, returns the URL that removes the credentials.

**Anything else you would like to add:**
[Additional information that will assist in solving the issue.]

_(see also: thread in [#carvel](https://kubernetes.slack.com/archives/CH8KCCKA5/p1619026581416200))_

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by pivotaljohn** [comment]
So, we've had a couple of runs at an implementation.

I'm realized that I can be more helpful if I focus my attention less on implementation details and more on API design.
The later is critical because we (rather aggressively) avoid deprecations in Carvel... and a key way to do that is spend the time really gaming out what the _end state_ API looks like (i.e. design broadly) and then implement just what we need right now (i.e. implement narrowly).

I've looked over:
- the Python standard library (specifically: [urllib](https://docs.python.org/3.8/library/urllib.parse.html#url-parsing))
- the Starlark spec (particularly the [dot notation bit](https://github.com/google/starlark-go/blob/master/doc/spec.md#dot-expressions))
- the Go standard library (as you have already: [net.url.URL](https://golang.org/pkg/net/url/#URL)

... for inspiration about what would be an intuitive and elegant API, here.

What comes to mind for me is something like this (stealing the testing bits you've already written):

```
url_type:
  #@ testURL1 = url.parse("http://admin:password@url.domain:8080")
  #@ testURL2 = testURL1.without_user()
  test1: #@ testURL1
  test2: #@ testURL1.user
  test3: #@ testURL1.user.name
  test4: #@ testURL1.user.password
  test5: #@ testURL2
  test6: #@ testURL2.user.name
  test7: #@ testURL2.user.password

+++

url_type:
  test1: http://admin:password@url.domain:8080
  test2: admin:password
  test3: admin
  test4: password
  test5: http://url.domain:8080
  test6: ""
  test7: ""
  test8: ""
```

Noting:
- values are accessed via the "dot operator"
- the shape of the structure follows the Go library designer's read of the relevant RFC: that `user` is a separate field.
- when a value isn't set, it's not `None`, but rather the falsey value of `""`

...

I'm gonna pause here.

@hanlins, two things for you:
- I apologize for not recognizing earlier that we needed to have a firmer set of "requirements" to avoid churn like this.
- What do you think about the design sketched out above?

**Comment by hanlins** [comment]
Hi @pivotaljohn thanks for the feedback, I think the dot representation makes sense and it seems more intuitive. I like this idea. I think the only thing I would argue is the results for empty values. For instance, if there's no user information in the string, then I would expect `null` instead of `""` as the result.
I've updated my PR, PTAL.

**Comment by pivotaljohn** [comment]
I think we can sneak in error messages instead of panics... reopening to attach that PR... if it's delayed at all, we can incorporate that into later work.

**Comment by pivotaljohn** [comment]
shipped with [v0.33.0](https://github.com/vmware-tanzu/carvel-ytt/releases/tag/v0.33.0).
